arcgis-ios-sdk-samples/Route & Directions/Navigate route/NavigateRouteViewController.swift

```diff
@@ -0,0 +1,296 @@
+                _ = routeTask
+            }
+        }
+    }
+    
+    /// Sets a speech synthesizer to generate voice guidance based on text.
+    func setSpeechSynthesizer() {
+        speechSynthesizer = AVSpeechSynthesizer()
+    }
+    
+    /// Sets the simulated data source for this demo.
+    /// - Parameter result: solved route from the route task.
+    func setDataSource(_ result: AGSRouteResult) {
+        // datasource should not be time-dependent
+        if let route = result.routes.first {
+            directionsList = route.directionManeuvers
+            let densifiedRoute = AGSGeometryEngine.geodeticDensifyGeometry(route.routeGeometry!, maxSegmentLength: 50.0, lengthUnit: .meters(), curveType: .geodesic) as! AGSPolyline
+            mockDataSource = AGSSimulatedLocationDataSource()
+            mockDataSource!.setLocationsWith(densifiedRoute)
+            mockDataSource!.locationChangeHandlerDelegate = self
+            mapView.locationDisplay.dataSource = mockDataSource!
+        }
+    }
+    
+    /// Sets the route tracker to provide navigation information.
+    /// - Parameter result: solved route from the route task.
+    func setRouteTracker(_ result: AGSRouteResult) {
+        routeTracker = AGSRouteTracker(routeResult: result, routeIndex: 0)
+        routeTracker!.delegate = self
+    }
+    
+    /// Sets the graphics.
+    /// - Parameter result: solved route from the route task.
+    func setRouteGraphics(_ result: AGSRouteResult) {
+        if let route = result.routes.first {
+            let stopSymbol = AGSSimpleMarkerSymbol(style: .diamond, color: .orange, size: 20)
+            for stop in makeStops() {
+                graphicsOverlay.graphics.add(AGSGraphic(geometry: stop.geometry, symbol: stopSymbol))
+            }
+            // Show the resulting route on the map and save a reference to the route.
+            routeAheadGraphic.geometry = route.routeGeometry
+            routeTraveledGraphic.geometry = nil
+            graphicsOverlay.graphics.addObjects(from: [
+                routeAheadGraphic,
+                routeTraveledGraphic
+            ])
+            if let routeGeometry = route.routeGeometry {
+                if let viewPoint = defaultViewPoint {
+                    // Reset to initial view point with animation.
+                    mapView.setViewpoint(viewPoint, completion: nil)
+                } else {
+                    mapView.setViewpointGeometry(routeGeometry) { [weak self] _ in
+                        // Get the initial zoomed view point.
+                        self?.defaultViewPoint = self?.mapView.currentViewpoint(with: .centerAndScale)
+                    }
+                }
+            }
+        }
+    }
+    
+    /// Resets to the starting location for location display.
+    func resetToStartingLocation() {
+        guard let initialLocation = mockDataSource?.locations?[0] else { return }
+        mockDataSource?.didUpdate(initialLocation)
+    }
```

Comment:
A few thoughts:

- This method is only being called from `reset()` and it doesn't really encapsulate anything, so I would just move the body of it into `reset()`.
- `if let` is being used to unwrap an optional, so there's no need to subscript into the `locations` array to get the first member. Just use `first` instead.
- `mockDataSource` is being unwrapped twice.

Putting those all together, this is what it might look like in `reset()`:

```swift
if let dataSource = mockDataSource, let location = dataSource.locations?.first {
    dataSource.didUpdate(location)
}
```